Support conditional bundling of assets based on `--flavor`

andrewkolos opened the pull request:
Provides support for conditional bundling of assets through the existing `--flavor` option for `flutter build` and `flutter run`. Closes https://github.com/flutter/flutter/issues/21682. Resolves https://github.com/flutter/flutter/issues/136092

## Change
Within the `assets` section pubspec.yaml, the user can now specify one or more `flavors` that an asset belongs to. Consider this example:

```yaml
# pubspec.yaml
flutter:
  assets:
    - assets/normal-asset.png
    - path: assets/vanilla/ice-cream.png
      flavors: 
        - vanilla
    - path: assets/strawberry/ice-cream.png
      flavors:
        - strawberry
```

With this pubspec,
* `flutter run --flavor vanilla` will not include `assets/strawberry/ice-cream.png` in the build output.
* `flutter run --flavor strawberry` will not include `assets/vanilla/ice-cream.png`.
* `flutter run` will only include `assets/normal-asset.png`.

## Open questions

* Should this be supported for all platforms, or should this change be limited to ones with documented `--flavor` support (Android, iOS, and (implicitly) MacOS)? This PR currently only enables this feature for officially supported platforms.

## Design thoughts, what this PR does not do, etc.

### This does not provide an automatic mapping/resolution of asset keys/paths to others based on flavor at runtime.

The implementation in this PR represents a simplest approach. Notably, it does not give Flutter the ability to dynamically choose an asset based on flavor using a single asset key. For example, one can't use `Image.asset('config.json')` to dynamically choose between different "flavors" of `config.json` (such as `dev-flavor/config.json` or `prod-flavor/config.json`). However, a user could always implement such a mechanism in their project or in a library by examining the flavor at runtime.

### When multiple entries affect the same file and 1) at least one of these entries have a `flavors` list provided and 2) these lists are not equivalent, we always consider the manifest to be ambiguous and will throw a `ToolExit`. 

<details>
For example, these manifests would all be considered ambiguous:

```yaml
assets:
  - assets/
  - path: assets/vanilla.png
    flavors: 
      - vanilla

assets:
  - path: assets/vanilla/
    flavors: 
      - vanilla
  - path: assets/vanilla/cherry.png
     flavor:
      - cherry

# Thinking towards the future where we might add glob/regex support and more conditions other than flavor:
assets:
  - path: assets/vanilla/**
    flavors:
      - vanilla
  - path: assets/**/ios/**
    platforms: 
       - ios

# Ambiguous in the case of assets like "assets/vanilla/ios/icon.svg" since we 
# don't know if flavor `vanilla` and platform `ios` should be combined using or-logic or and-logic.
```

See [this review comment thread](https://github.com/flutter/flutter/pull/132985#discussion_r1381909942) for the full story on how I arrived at this decision.
</details>


### This does not support Android's multidimensional flavors feature (in an intuitive way)

<details>

Conder this excerpt from a Flutter project's android/app/build.gradle file:

```groovy
android {
    // ...

    flavorDimensions "mode", "api"

    productFlavors {
        free {
            dimension "mode"
            applicationIdSuffix ".free"
        }

        premium {
            dimension "mode"
            applicationIdSuffix ".premium"
        }

        minApi23 {
            dimension "api"
            versionNameSuffix "-minApi23"
        }

        minApi21 {
            dimension "api"
            versionNameSuffix "-minApi21"
        }
    }
}
```

In this setup, the following values are valid `--flavor` are valid `freeMinApi21`, `freeMinApi23`, `premiumMinApi21`, and `premiumMinApi23`. We call these values "flavor combinations". Consider the following from the Android documentation[^1]:

> In addition to the source set directories you can create for each individual product flavor and build variant, you can also create source set directories for each combination of product flavors. For example, you can create and add Java sources to the src/demoMinApi24/java/ directory, and Gradle uses those sources only when building a variant that combines those two product flavors.
> 
> Source sets you create for product flavor combinations have a higher priority than source sets that belong to each individual product flavor. To learn more about source sets and how Gradle merges resources, read the section about how to [create source sets](https://developer.android.com/build/build-variants#sourcesets).

This feature will not behave in this way. If a user utilizes this feature and also Android's multidimensional flavors feature, they will have to list out all flavor combinations that contain the flavor they want to limit an asset to:

```yaml
assets:
  - assets/free/
    flavors:
      - freeMinApi21
      - freeMinApi23
```

This is mostly due to a technical limitation in the hot-reload feature of `flutter run`. During a hot reload, the tool will try to update the asset bundle on the device, but the tool does not know the flavors contained within the flavor combination (that the user passes to `--flavor`). Gradle is the source of truth of what flavors were involved in the build, and `flutter run` currently does not access to that information since it's an implementation detail of the build process. We could bubble up this information, but it would require a nontrivial amount of engineering work, and it's unclear how desired this functionality is. It might not be worth implementing.

</details>



See https://flutter.dev/go/flavor-specific-assets for the (outdated) design document. 

<summary>Pre-launch Checklist</summary>

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x]  I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

</details>

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

[^1]: https://developer.android.com/build/build-variants#flavor-dimensions

andrewkolos committed:
squash

oops

enable for `flutter daemon` and `flutter run --machine`

nit

working (ugly) flavor-collision-handling logic

nits

add commented out tests

refactor main test a little

update tests and add support for deferred components assets

remove extra log statements

add a test for assetsentry parsing

get build_info.dart changes under test

fix asset bundle test for windows

make asset bundle tests work on Windows

typos

remove feature for web

typo

fix some tests

fix xcode_backend_test

remove feature for web

fix a test

fix oopsie

fix test

try fix xcode backend test

restore in-order sequencing of install tests

augment flavors integration tests with flavor-specific assets checks

add support to HotRunner (untested)

more diff cleanup

remove support for flavors for platforms that we haven't documented flavors support for

enable for web

add tests to asset bundle tests

add tests for CopyAssets target and fix logic for unflavored builds

clean up print statements

simplify pubspec syntax

implement group pubspec syntax (very hacky)

group parsing actually works (image displays at runtime)

xcode sends flavor, we use flavor in asset bundling (hacky)

gradle-sends-flavor-to-flutter-assembler

andrewkolos committed:
cleanup

jonahwilliams commented on packages/flutter_tools/bin/xcode_backend.dart:
```diff
@@ -409,11 +409,16 @@ class Context {
       '-dDartObfuscation=${environment['DART_OBFUSCATION'] ?? ''}',
       '-dAction=${environment['ACTION'] ?? ''}',
       '-dFrontendServerStarterPath=${environment['FRONTEND_SERVER_STARTER_PATH'] ?? ''}',
+      '-dFlavor=${environment['FLAVOR'] ?? ''}',
       '--ExtraGenSnapshotOptions=${environment['EXTRA_GEN_SNAPSHOT_OPTIONS'] ?? ''}',
       '--DartDefines=${environment['DART_DEFINES'] ?? ''}',
       '--ExtraFrontEndOptions=${environment['EXTRA_FRONT_END_OPTIONS'] ?? ''}',
     ]);
 
+    if (environment['FLAVOR'] != null && environment['FLAVOR']!.isNotEmpty) {
+      flutterArgs.add('-dFlavor=${environment['FLAVOR'] ?? ''}');
+    }
```
Looks like you're adding this twice?

jonahwilliams commented on packages/flutter_tools/lib/src/asset.dart:
```diff
@@ -1032,16 +1111,22 @@ class ManifestAssetBundle implements AssetBundle {
 
+  /// The platform-independent URL provided by the user in the pubspec that this
+  /// asset was found from.
+  final Uri originUri;
```
What is this being used for?

andrewkolos replied:
See https://github.com/flutter/flutter/pull/132985#discussion_r1383401264. In short, it's to help disambiguate in scenarios where multiple declarations affect the same asset:

```yaml
assets:
  - path: assets/paid/
    flavor: paid
  - path: assets/paid/pip.mp3 # For pip.mp3, this declaration overrides the previous one.
    flavor: free
```

This could probably just be a `ToolExit` instead though.

jonahwilliams commented on packages/flutter_tools/lib/src/asset.dart:
```diff
@@ -938,10 +969,48 @@ class ManifestAssetBundle implements AssetBundle {
+      flavor: flavor,
     );
+
+    final bool hasMoreSpecificDeclarationWithFlavor = ((){
+      final _Asset? preExistingAsset = result.keys
+        .where((_Asset other) => other.entryUri == asset.entryUri)
+        .firstOrNull;
+      if (preExistingAsset == null) {
+        return false;
+      }
+
+      if (flavor != null && flavor != preExistingAsset.flavor ) {
+        if (assetUri.path.endsWith('/') == preExistingAsset.originUri.path.endsWith('/')) {
+          throwToolExit('Asset "${assetUri.path}" belongs to multiple flavors.\n'
+            'An entry with the path "${preExistingAsset.originUri.path}" gives it a flavor of "${preExistingAsset.flavor}".\n'
+            'An entry with the path "${assetUri.path}" gives it a flavor of "$flavor".'
+          );
+        }
+        if (assetUri.path.endsWith('/') && !preExistingAsset.originUri.path.endsWith('/')) {
+          return true;
+        }
+        else if (!assetUri.path.endsWith('/') && preExistingAsset.originUri.path.endsWith('/')) {
+          _logger.printTrace('Skipping assets entry "${asset.entryUri.path}" since its configured flavor, "${asset.flavor}", did not match the provided flavor (if any).');
+          result.remove(preExistingAsset);
+        }
+      }
+
+      return false;
+    })();
```
Rather than an IIFE, please pull this into a helper method.

jonahwilliams commented on packages/flutter_tools/lib/src/asset.dart:
```diff
@@ -938,10 +969,48 @@ class ManifestAssetBundle implements AssetBundle {
     final File assetFile = asset.lookupAssetFile(_fileSystem);
-    for (final String path in cache.variantsFor(assetFile.path)) {
+
+    final List<String> relevantFiles =  cache.variantsFor(assetFile.path);
+
+    for (final String path in relevantFiles) {
```
?

andrewkolos replied:
Indeed. This was probably an artifact of some earlier experiment. Fixed in [7ae8ba2](https://github.com/flutter/flutter/pull/132985/commits/7ae8ba2f8e304ae81dcb8a72a927200d12d87de2)

jonahwilliams commented on packages/flutter_tools/lib/src/asset.dart:
```diff
@@ -938,10 +969,48 @@ class ManifestAssetBundle implements AssetBundle {
+        return false;
+      }
+
+      if (flavor != null && flavor != preExistingAsset.flavor ) {
+        if (assetUri.path.endsWith('/') == preExistingAsset.originUri.path.endsWith('/')) {
+          throwToolExit('Asset "${assetUri.path}" belongs to multiple flavors.\n'
+            'An entry with the path "${preExistingAsset.originUri.path}" gives it a flavor of "${preExistingAsset.flavor}".\n'
+            'An entry with the path "${assetUri.path}" gives it a flavor of "$flavor".'
+          );
```
Can you ellaborate a bit on what this error state is? Its an error in your implementation to have a flavor specific asset with more than one flavor by listing it twice?

andrewkolos replied:
I've tried to capture the intended behavior in the [tests](https://github.com/flutter/flutter/blob/793eab617bdb8420243c84f3054882a0e6efdbd6/packages/flutter_tools/test/general.shard/asset_bundle_test.dart#L335). In short, the answer is, yes, this implementation doesn't support multiple flavors for one asset. However, we could add support for that by, for example, changing the manifest syntax to accept a `flavors` list rather than a `flavor` string.

The error logic you pointed out is certainly complicated. To better understand where it came from, start by considering this case:

```yaml
assets:
  - assets/
  - path: assets/free.mp3
    flavor: free
```

To me, it is intuitive to bundle `free.mp3` if and only if `--flavor` is set to `free`. Following that logic, my first intuition was that any asset covered by an entry with a `flavor` is globally limited to that flavor. Therefore, two declarations that list the same asset but with different flavors results in an ambiguous manifest[^1]:
```yaml
assets:
  - path: assets/pip.mp3
    flavor: free
  - path: assets/pip.mp3
    flavor: paid

# asset/pip.mp3 can't be limited *only* to "free" and also *only* to "paid"
```

To finally circle back to the weird error logic you highlighted, consider this kind of case (which is hopefully extremely rare):

```yaml
assets:
  - path: assets/paid/
    flavor: paid
  - path: assets/paid/pip.mp3
    flavor: free
```

Here, for `pip.mp3`, the second entry takes precedence over the first since it is more specific. However, in hindsight, I think it's okay to make the code simpler by having this also result in a `ToolExit`, because I don't expect many—if any—users would specify something like this. 

One alternative to all of this is reconsidering one of my old ideas where flavor-specific assets aren't listed in the asset manifest but rather within separate flavor manifests:
```yaml
assets:
  - assets/
# ...
flavors:
  - name: free
    additional-assets:
     - assets/free.mp3
```

Still, if we add other conditional logic to assets in the future (e.g. platform-specificity), we'll still run into the same kind of problem. What if I have an asset that is specific to a combination of platform and flavor?

[^1]: Now that you point it out though, it's possible some users may intuit that it's possible to give an asset multiple flavors by listing it multiple times with different flavors (I'm guessing this would also result in a simpler implementation).

jonahwilliams replied:
Interesting! As a side note, these sort of comments are absolute 🥇 so consider adding these as documentation in the code.

I don't remember, but I suspect what happens today is that when you list the same asset multiple times, or list a directory and then a specific asset, that it gets ignored? And I suppose this probably works fine since we only need a signal as to whether or not the asset is included - there is no additional metadata.

But now with flavor specific assets, we do have additional data. I think for the first version I like this approach - lets keep it simple and define this as an error case, we can always weaken it later. The alternative is having rules for resolving confilicts which sounds like it would only be more complicated.

Consider pulling this code into its own method with your comment as a docstring.

andrewkolos replied:
> I don't remember, but I suspect what happens today is that when you list the same asset multiple times, or list a directory and then a specific asset, that it gets ignored?

IIRC, redundant entries will get processed, but (as you alluded to) it's essentially a no-op (we just overwrite entries in the output manifest with identical data).

> But now with flavor specific assets, we do have additional data. I think for the first version I like this approach - lets keep it simple and define this as an error case, we can always weaken it later. 

I assume you are referring to this case (and implicitly the one just before it):

```yaml
assets:
  - path: assets/paid/
    flavor: paid
  - path: assets/paid/pip.mp3
    flavor: free
```

If so, yeah, I think it makes sense to just have this be an error case for now. Initially, I thought this case might be more common if/when we support glob/regex declarations, but even then I don't think any significant number of users would need conflict-resolving logic.

> Consider pulling this code into its own method with your comment as a docstring.

Sure, I'll work on that. I'm glad you enjoyed my writing :)

christopherfujino replied:
Everything here SGTM

andrewkolos replied:
I did some more pondering on this (cheers to @jonahwilliams for chatting with me on this). To recap, the central question is "what should we do when multiple asset entries without identical[^1] conditions affect the same file?" Should conditions from separate entries be considered disjunctively (with "or" logic) or conjunctively (with "and" logic)? Should it depend on the specificity of those entries (i.e. file vs directory)? I don't believe there are obvious answers to these questions, but we don't need answers right now—if at all. We can instead throw a `ToolExit` when such ambiguities arise, and if users later overwhelmingly agree that such conditions should be combined in a specific way, we can change the behavior later without breaking anyone. In the meantime, the users can avoid these `ToolExit`s by writing more explicit `assets` manifests and/or organizing assets into individual directories[^2].

To be clear, all the example `assets` manifests I provided previously in this thread (including the first one) would result in a `ToolExit` being thrown. Here's a specific example to focus on:

```yaml
assets:
  - assets/
  - path: assets/free.mp3
    flavor: free
```

Assuming the file `assets/free.mp3` exists, this would result in a `ToolExit`. The message would explain that the tool could not determine whether or not to bundle `assets/free.mp3` and where the ambiguity the came from, along with a vague—but still useful—hint as to how to remove the ambiguity.

[^1]: To be clear, an entry _without_ any explicit conditions and an entry _with_ conditions are still considered to not have identical conditions.
[^2]: Users would most likely already do this today, but they might get more creative if/when we support glob/regex patterns for asset declaration.

jonahwilliams commented on packages/flutter_tools/lib/src/commands/assemble.dart:
```diff
@@ -116,6 +116,7 @@ class AssembleCommand extends FlutterCommand {
         'files will be written. Must be either absolute or relative from the '
         'root of the current Flutter project.',
     );
+    usesFlavorOption();
```
So this adds the --flavor command line flag, but you're passing the flavor with -D option which is separate.

jonahwilliams commented on packages/flutter_tools/lib/src/flutter_manifest.dart:
```diff
@@ -640,17 +596,48 @@ void _validateDeferredComponents(MapEntry<Object?, Object?> kvp, List<String> er
+  return assetsErrors.isEmpty;
+}
+
+List<AssetsEntry> _computeAssetsSafe(Object? yaml, List<String> errors) {
```
consider returning a tuple of assets, errors instead of passing in the error list separately

andrewkolos replied:
Resolved in [c729929](https://github.com/flutter/flutter/pull/132985/commits/c72992997c2d24001eb81755301b32ced9694ad4)

(courtesy reminder that you can any ignore email notifications from comments—I'll re-request review when ready)

jonahwilliams commented on packages/flutter_tools/lib/src/flutter_manifest.dart:
```diff
@@ -703,3 +690,97 @@ void _validateFonts(YamlList fonts, List<String> errors) {
+  }
+
+  @override
+  bool operator ==(Object other) {
+    if (other is! AssetsEntry) {
+      return false;
+    }
+
+    if (uri != other.uri) {
+        return false;
+      }
+
+    return true;
+  }
+
+  @override
+  int get hashCode => uri.hashCode;
```
Is it correct that this ignores flavor?

andrewkolos committed:
remove duplicate Xcode define

andrewkolos committed:
revert unnecessary extraction of expression into local variable

andrewkolos committed:
make AssetsEntry parsing functions pure

andrewkolos committed:
remove redundant --flavor option in assemble command

andrewkolos committed:
if flavor is null or empty, do not pass it to xcode via -d

andrewkolos committed:
allow empty flavor when passing to xcode

This shouldn't really matter because the app won't shouldn't build with any empty flavor.

christopherfujino commented on packages/flutter_tools/lib/src/asset.dart:
```diff
@@ -621,39 +625,50 @@ class ManifestAssetBundle implements AssetBundle {
-          wildcardDirectories.add(assetUri);
+      final Map<_Asset, List<_Asset>> componentAssets = <_Asset, List<_Asset>>{};
+      for (final AssetsEntry assetsEntry in component.assets) {
+        if (assetsEntry.flavor != null && assetsEntry.flavor != flavor) {
```
I suggest ignoring the case where `flavor == ''` since we default to empty string

```suggestion
        if (assetsEntry.flavor != null && flavor != '' && assetsEntry.flavor != flavor) {
```

andrewkolos replied:
I've removed the defaulting, so this shouldn't be a problem anymore.

andrewkolos committed:
disallow overlapping entries of different flavors

andrewkolos committed:
fix flavors for deferred components

andrewkolos committed:
fix some tests

andrewkolos committed:
consider flavor in AssetsEntry.hashCode

jonahwilliams commented on packages/flutter_tools/lib/src/asset.dart:
```diff
@@ -963,13 +1004,73 @@ class ManifestAssetBundle implements AssetBundle {
+      return;
+    }
+
+    String errorMessage =
```
nit: use a [StringBuffer](https://api.dart.dev/stable/3.2.0/dart-core/StringBuffer-class.html)

andrewkolos replied:
I checked with #hackers-dart, and I learned that the Dart compiler [does not automatically optimize this in](https://discord.com/channels/608014603317936148/608022273152122881/1174413326231543879). I've replaced `errorMessage` with a `StringBuffer` in [c494656](https://github.com/flutter/flutter/pull/132985/commits/c494656285fcc3e9406f2f61335532c6f46d8395#diff-1339eded5eeb35ab9fbc61051c530c73e307d4640328760c1af0eee36c142be2R1056).

andrewkolos replied:
https://github.com/dart-lang/sdk/issues/50304#issuecomment-1291705018

jonahwilliams commented on packages/flutter_tools/lib/src/flutter_manifest.dart:
```diff
@@ -703,3 +691,88 @@ void _validateFonts(YamlList fonts, List<String> errors) {
+
+    if (yaml is String) {
+      final (Uri? uri, String? error) = tryParseUri(yaml);
+      return uri == null ? (null, error) : (AssetsEntry(uri: uri), null);
+    } else if (yaml is Map) {
```
If you return you don't need to do an else:

```dart
if (yaml is String) {
  ...
  return ..;
}

if (yaml is Map) {
  ...
}
```

andrewkolos replied:
Addressed in [nits](https://github.com/flutter/flutter/pull/132985/commits/961903a75eb72d4a9b92ee8c51c3bd9ac3d7c8b8)

jonahwilliams commented on packages/flutter_tools/lib/src/flutter_manifest.dart:
```diff
@@ -703,3 +691,88 @@ void _validateFonts(YamlList fonts, List<String> errors) {
+      }
+
+      if (flavor != null && flavor is! String) {
+        return( null, 'Asset manifest entry is malformed. '
```
```suggestion
        return (null, 'Asset manifest entry is malformed. '
```

jonahwilliams commented on packages/flutter_tools/lib/src/flutter_manifest.dart:
```diff
@@ -703,3 +691,88 @@ void _validateFonts(YamlList fonts, List<String> errors) {
+  final Uri uri;
+  final String? flavor;
+
+  static const String _pathKey = 'path';
```
constants should have a `k` prefix: `static const String _kPathKey = 'path'`

andrewkolos replied:
IIRC our style guide only applies this rule to global constants ([wiki](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#begin-global-constant-names-with-prefix-k)). Do you believe all constants should be prefixed?

jonahwilliams replied:
I generally do, but perhaps we don't follow this as strictly as I thought.

jonahwilliams commented on packages/flutter_tools/lib/src/flutter_manifest.dart:
```diff
@@ -703,3 +691,88 @@ void _validateFonts(YamlList fonts, List<String> errors) {
+    if (other is! AssetsEntry) {
+      return false;
+    }
+
+    if (uri != other.uri || flavor != other.flavor) {
+        return false;
+    }
+
+    return true;
+  }
```
```suggestion

    return uri == other.uri && flavor == other.flavor;
  }
```

andrewkolos replied:
Addressed in [nits](https://github.com/flutter/flutter/pull/132985/commits/961903a75eb72d4a9b92ee8c51c3bd9ac3d7c8b8)

jonahwilliams commented on packages/flutter_tools/lib/src/run_hot.dart:
```diff
@@ -96,6 +96,7 @@ class HotRunner extends ResidentRunner {
     ReassembleHelper reassembleHelper = _defaultReassembleHelper,
     NativeAssetsBuildRunner? buildRunner,
     required Analytics analytics,
+    this.flavor,
```
Consider putting flavor on debugging options instead.

andrewkolos replied:
This reminded me that the flavor is already available in `debuggingOptions.buildInfo`, so we can just reference that. Addressed in [remove redundant flavor arg from HotRunner](https://github.com/flutter/flutter/pull/132985/commits/03550d0dd9c19673799e0c813c8b129d7fd9fdc9)

jonahwilliams reviewed (commented):
Change seems ready overall, what else needs to be done?

andrewkolos commented:
> Change seems ready overall, what else needs to be done?

I'm changing `flavor` to `flavors` and allowing multiple flavor values. After I do that and update some tests, this will be ready for another review pass.

andrewkolos committed:
change `flavor` to `flavors` to allow multiple flavors per asset

andrewkolos committed:
nits

andrewkolos committed:
remove redundant flavor arg from HotRunner

andrewkolos committed:
oops

andrewkolos committed:
oops again

christopherfujino commented on packages/flutter_tools/lib/src/asset.dart:
```diff
@@ -621,39 +625,49 @@ class ManifestAssetBundle implements AssetBundle {
+
+      componentAssets.removeWhere((_Asset asset, List<_Asset> variants) {
+        if (!asset.matchesFlavor(flavor)) {
+          _logger.printTrace('Skipping assets entry "${asset.entryUri.path}" since '
+            'its configured flavor(s) did not match the provided flavor (if any).\n'
+            'Configured flavors: ${asset.flavors.join(', ')}\n');
```
do we need this?

andrewkolos replied:
I rely on this to see if an asset got bundled by `flutter build` in integration tests ([code](https://github.com/flutter/flutter/pull/132985/files#diff-53383b32b975bfed6875306dfb98911cad077a5251ca0591c5b0e125fb4a0f05R49)). I am not a big fan of depending on logs for testing, but I think the other way to do this would be to delve into the output of `flutter build` and look for asset files there, which might require a significant amount of effort to implement 🤷 .

christopherfujino replied:
this will get written to AssetManifest.json, right?

andrewkolos replied:
Yeah, one could decode the AssetManifest and effectively get a list of assets that were bundled

christopherfujino replied:
ok, let me try to modify this test to add that

andrewkolos replied:
> ok, let me try to modify this test to add that

@christopherfujino IIRC, you have a branch that does this now. Feel free to push merge it into this branch or send a link to it, and I can integrate it

christopherfujino replied:
thanks for the ping, opened https://github.com/andrewkolos/flutter/pull/1

christopherfujino replied:
didn't run it on CI, but it passes for me locally

christopherfujino committed:
test based on android manifest (#1)

* remove trace and fix test

* update to use AssetManifest.bin; mirror change in flavors_test_ios

---------

Co-authored-by: Andrew Kolos <andrewrkolos@gmail.com>

andrewkolos commented:
@jonahwilliams @christopherfujino friendly ping

jonahwilliams reviewed (approved):
LGTM (but not to CI)

gnprice commented:
Neat!

Is there documentation written down somewhere for this feature? I feel like having documentation will be a key part of making this a feature that's fully available to Flutter users. It's also helpful to have written the docs before merging, because sometimes there's a bit that feels hard to explain and that can point to a way to make the API better.

For most kinds of Flutter features, the documentation gets written naturally alongside the code, in dartdoc comments. For this feature that doesn't apply, and I suppose somewhere on https://docs.flutter.dev/ui/assets/assets-and-images is the right home for docs. So a parallel PR for the website, adding the docs alongside this PR adding the implementation, would be excellent.

andrewkolos commented:
> Is there documentation written down somewhere for this feature?

Not yet.

> I feel like having documentation will be a key part of making this a feature that's fully available to Flutter users. It's also helpful to have written the docs before merging, because sometimes there's a bit that feels hard to explain and that can point to a way to make the API better.

I agree. Thanks for the reminder! I'll get started on a draft right away.

christopherfujino commented:
what's the plan with handling analysis?

andrewkolos commented:
> what's the plan with handling analysis?

Nothing concrete yet. From https://github.com/dart-lang/sdk/issues/54038, I'm assuming @srawlins is going to implement the needed analyzer changes, but I'll follow up with him. I don't plan on merging this until the needed analyzer changes have rolled in, and I'm guessing there's plenty of time until the next major release.

christopherfujino commented:
> > what's the plan with handling analysis?
> 
> Nothing concrete yet. From [dart-lang/sdk#54038](https://github.com/dart-lang/sdk/issues/54038), I'm assuming @srawlins is going to implement the needed analyzer changes, but I'll follow up with him. I don't plan on merging this until the needed analyzer changes have rolled in, and I'm guessing there's plenty of time until the next major release.

My concern is less missing the feature cut and more the toil of dealing with upstream breaking changes.

christopherfujino commented on packages/flutter_tools/lib/src/asset.dart:
```diff
@@ -1032,16 +1140,22 @@ class ManifestAssetBundle implements AssetBundle {
 
   final Package? package;
 
+  /// The platform-independent URL provided by the user in the pubspec that this
```
this is a URI, not a URL, right?

gnprice replied:
FWIW that distinction is deprecated in the modern web standards, and just "URL" is now the standard term: https://url.spec.whatwg.org/#goals

It looks like in Flutter docs there's already somewhat more use of "URL" than "URI". So I'd suggest going ahead with "URL" in newly-written text.

(A pair of searches, a bit complicated in order to try to match just when it's a word, adapted from searches I've done elsewhere for a similar purpose:
```
$ git grep '///.*(Url(\b|[_A-Z])|(\b|_)url(\b|[_A-Z])|URL)' packages/*/lib | wc -l
177
$ git grep '///.*(Uri(\b|[_A-Z])|(\b|_)uri(\b|[_A-Z])|URI)' packages/*/lib | wc -l
71
```
)

christopherfujino commented on packages/flutter_tools/lib/src/asset.dart:
```diff
@@ -1065,6 +1181,21 @@ class _Asset {
+      return true;
+    }
+
+    return flavors.any((String e) => e == flavor);
```
today it matches nothing. which is better than matching everything, assuming we reached this line because `flavors.isNotEmpty`. is it valid to have flavors configured on a project and not pass one? and if not, how do we handle that case?

andrewkolos replied:
> is it valid to have flavors configured on a project and not pass one?

Great question. If I had to guess, the behavior for this hasn't been defined. Suppose you have a Flutter project with a `free` flavor configured. If you try to run/build without `--flavor` on Android, you'll see this error similar to this:
```
Gradle build failed to produce an .apk file. It's likely that this file was generated under /Users/andrewkolos/Desktop/deleteme/throwaway/build, but the tool couldn't find it.
```

However, you can build/run against iOS without `--flavor` without issue. 🤷 

To return on the root question:

> Also what does a null flavor imply. Should that match everything?

Yes, it's intended that a flavorless asset matches any flavor, including the lack of a flavor (`null`). I've made this more explicit by adding another early return. https://github.com/flutter/flutter/blob/f7e61ebf295dc530a8199d220bb99f03c1a9ad19/packages/flutter_tools/lib/src/asset.dart#L1184-L1194

christopherfujino commented on packages/flutter_tools/lib/src/asset.dart:
```diff
@@ -1065,6 +1181,21 @@ class _Asset {
+  bool hasEquivalentFlavorsWith(_Asset other) {
+    final Set<String> assetFlavors = flavors.toSet();
+    final Set<String> otherFlavors = other.flavors.toSet();
+    final Set<String> union = assetFlavors.union(otherFlavors);
+    return union.every((String e) => assetFlavors.contains(e) && otherFlavors.contains(e));
```
```suggestion
    return assetFlavors.length == otherFlavors.length && assetFlavors.every(
      (String e) => otherFlavors.contains(e),
    );
```

christopherfujino commented on packages/flutter_tools/lib/src/asset.dart:
```diff
@@ -1080,7 +1211,8 @@ class _Asset {
         && other.relativeUri == relativeUri
         && other.entryUri == entryUri
-        && other.assetKind == assetKind;
+        && other.assetKind == assetKind
```
should we add these fields to the hashcode?

christopherfujino replied:
`Object.hashAll()` will take an interable, so we could do:

```dart
  @override
  int get hashCode => Object.hashAll(<Object>[
    baseDir,
    relativeUri,
    entryUri,
    assetKind,
    ...flavors,
  ]);
```

andrewkolos committed:
nit

Co-authored-by: Christopher Fujino <fujino@google.com>

christopherfujino commented on packages/flutter_tools/lib/src/base/deferred_component.dart:
```diff
@@ -95,8 +96,11 @@ class DeferredComponent {
+    for (final AssetsEntry asset in assets) {
+      out.write('\n    - ${asset.uri.path}');
+      if (asset.flavors.isNotEmpty) {
+        out.write(' (flavor: ${asset.flavors})');
```
should this be "flavors", plural?

andrewkolos committed:
nits

christopherfujino commented on packages/flutter_tools/lib/src/commands/daemon.dart:
```diff
@@ -511,6 +511,7 @@ class AppDomain extends Domain {
     bool machine = true,
     String? userIdentifier,
     bool enableDevTools = true,
+    String? flavor,
```
do we need this?

andrewkolos replied:
No, we shouldn't. Flavor is already contained within `DebuggingOptions`. I removed this.

andrewkolos committed:
remove unused flavor parameter

christopherfujino commented on packages/flutter_tools/lib/src/flutter_manifest.dart:
```diff
@@ -640,17 +596,48 @@ void _validateDeferredComponents(MapEntry<Object?, Object?> kvp, List<String> er
+  return errors;
+}
+
+(List<AssetsEntry>, List<String> errors) _computeAssetsSafe(Object? yaml) {
```
if I'm reading this code correctly, When we call `FlutterManifest._createFromYaml()`, we will be parsing these assets twice, once from calling `_validate()` -> `_validateFlutter()` -> `_validateAssets()` -> `computeAssetsSafe()` and later, from `FlutterManifest.assets` -> `_computeAssets()` -> `computeAssetsSafe()`.

And if you have deferred components we will parse it many more times. Can we call this function *once* from `FlutterManifest._createFromYaml()` and re-use both the erros and the `List<AssetEntry>`?

andrewkolos replied:
Yes, this is correct, and this pattern is found throughout the `FlutterManifest` class (see the `fonts`, `shaders`, and `models` fields). I'd be happy to attempt to refactor this class to parse and store everything upfront, but this would be a substantial refactoring. Because of this, I elected not to break this pattern in this PR. Do you believe I should attempt this refactoring now?

andrewkolos replied:
@christopherfujino friendly ping

andrewkolos replied:
Done in [add tech debt note about redundant parsing](https://github.com/flutter/flutter/pull/132985/commits/00a5a87bc727274a037bdb4ab10f72a08c04aceb)

christopherfujino reviewed (approved):
LGTM

andrewkolos committed:
make flavor matching behavior more explicit

andrewkolos committed:
add tech debt note about redundant parsing

andrewkolos committed:
fixing rebasing oopsies

rebasing because frob asked me to

XilaiZhang commented:
Looks like Google Testing got stuck after clicking on the rerun button (filed https://b.corp.google.com/issues/315219598)

pushed a new commit to see if it unblocks the Google Testing.

If this is urgent just go ahead and override the Google Testing status

andrewkolos committed:
fix test: clean up after flavors test

XilaiZhang commented:
just deployed a new version of roll worker. @andrewkolos could you try resolving the conflict and push a new commit to trigger the build? thank you!

andrewkolos committed:
run update-packages

XilaiZhang commented:
Looks like Google Testing completed

christopherfujino commented:
Reverting this as it appears to have broken: [Mac_ios flavors_test_ios](https://ci.chromium.org/p/flutter/builders/prod/Mac_ios%20flavors_test_ios)

giaur500 commented:
That's not clear for me:

`Notably, it does not give Flutter the ability to dynamically choose an asset based on flavor using a single asset key. For example, one can't use Image.asset('config.json') to dynamically choose between different "flavors" of config.json (such as dev-flavor/config.json or prod-flavor/config.json). However, a user could always implement such a mechanism in their project or in a library by examining the flavor at runtime.`

Looking at you example:

```
# pubspec.yaml
flutter:
  assets:
    - assets/normal-asset.png
    - path: assets/vanilla/ice-cream.png
      flavors: 
        - vanilla
    - path: assets/strawberry/ice-cream.png
      flavors:
        - strawberry

```
Both flavors: `vanilla` and `strawberry` will include asset named `ice-cream.png`. So I use something like this:

`Image.asset('ice-cream.png')`

I will get image currently bundled (different for vanilla and different for strawberry flavor), is that right? Or I need to do something else?

andrewkolos commented:
Hello @giaur500,

> Both flavors: `vanilla` and `strawberry` will include asset named `ice-cream.png`, but file will be taken from different direcory (as defined in publspec). So I use something like this:
> 
> `Image.asset('ice-cream.png')`
> 
> I will get image currently bundled (different for vanilla and different for strawberry flavor), is that right? Or I need to do something else?

This PR _only_ changes what asset files get bundled with an app. It does _not_ change how `Image.asset` resolves asset names, so `'ice-cream.png'` would not resolve to any file, because it doesn't exist. Your app would have to use either `Image.asset('assets/strawberry/ice-cream.png')` or `Image.asset('assets/vanilla/ice-cream.png')`. If you would like a single name to resolve another name at runtime based on the flavor of the app, you would have to write this logic yourself. For example, here's one way you could do that:

```dart
import 'package:flutter/services.dart';

/// Takes an asset name and inserts the flavor directory into it.
///
/// For example, if the app flavor is "vanilla" then flavored("assets/ice-cream.png")
/// will resolve to "assets/vanilla/ice-cream.png".
///
/// Expects [assetName] to begin with "assets/". For example,
/// "assets/ice-cream.png".
String flavored(String assetName) {
  if (!assetName.startsWith('assets/')) {
    throw Exception('Flavored asset name must start with "assets/"');
  }
  if (appFlavor == null) {
    throw Exception('Cannot get flavor of an asset when the app flavor is null.');
  }

  final String assetNameWithoutAssetsDir = assetName.substring('assets/'.length);
  return 'assets/$appFlavor/$assetNameWithoutAssetsDir';
}

// ...
  
  // This will resolve to 'assets/strawberry/ice-cream.png' or 'assets/vanilla/ice-cream.png' based 
  // on the value of the global appFlavor variable at runtime.
  Image.asset(flavored('assets/ice-cream.png'));

// ... 
```

I considered having Flutter automatically map asset names to other names at runtime, but this would require more consideration in its design (which I would be happy to discuss further on a feature-request issue, if/when it's filed). I didn't want this PR to grow more complex than it already is.

pan9966 commented:
Ok, but I don't think it solves main problem, unless we have any way to determine what is current flavor at runtime. Is there any? What is `appFlavor` in your example?

Instead playing with flavor names at runtime, can we define assets in publspec, but bundle it with the same name (but, take it from subdirectory related to flavor). So asset will get always the same name, but different assets will be bundled.

giaur500 commented:
Agree, condional bundling should mean condional bundling the same assets. Similar as native Android resources - you can define separate dir for each flavor to bundle.

You can do conditional bundle assets, but you can't access them, because assets path vary. Also, there's no any api to determine flavor, so how to access assets bundled this way?

We need either consider some way to make possible to define multiple dirs in publspec, but don't bundle them as is - they should be bundled with the same path, or provide some way to determine current flavor at runtime.

andrewkolos commented:
> Ok, but I don't think it solves main problem, unless we have any way to determine what is current flavor at runtime. Is there any? What is `appFlavor` in your example?

Ah, sorry for including this information in my prior comment. [`appFlavor`](https://api.flutter.dev/flutter/services/appFlavor-constant.html) is part of the services API. It will contain whatever you set `--flavor` to when running/building your app. I've filed to https://github.com/flutter/website/issues/9913 to get this more visibly documented.

> Instead playing with flavor names at runtime, can we define assets in publspec, but bundle it with the same name (but, take it from subdirectory related to flavor). So asset will get always the same name, but different assets will be bundled.

I considered this type of approach, but this would have required much more design[^1] and implementation work. I wanted to get _some_ solution out more quickly, hence this approach.

[^1]: There's also a desire for more types of conditional bundling, see https://github.com/flutter/flutter/issues/96514, so I want a solution that scales up to many conditions in a usable way. I will try to find some more time to add more thoughts on that issue thread.

Gilianp commented:
Is this feature already available? No matter what syntax I use, none of it works for me. The asset is always not included in the final flavor build

giaur500 commented:
> Is this feature already available? No matter what syntax I use, none of it works for me. The asset is always not included in the final flavor build

Have tried master branch? I don't think it's merged to stable Flutter so far

Gilianp commented:
> > Is this feature already available? No matter what syntax I use, none of it works for me. The asset is always not included in the final flavor build
> 
> Have you tried master branch? I don't think it's merged to stable Flutter so far

Sorry for delay. I always use master branch

gnprice commented:
@Gilianp If you're still finding this feature doesn't work for you, please file a new issue and include all the details requested in the issue template. That will make it possible to investigate the cause of what you're seeing.

Gilianp commented:
> @Gilianp If you're still finding this feature doesn't work for you, please file a new issue and include all the details requested in the issue template. That will make it possible to investigate the cause of what you're seeing.

Ok @gnprice. I will prepare a new issue with all details. Thanks

giaur500 commented:
@andrewkolos Is there any example app to build and try?

andrewkolos commented:
> @andrewkolos Is there any example app to build and try?

Sure, you can use this project with `--flavor vanilla`:
https://github.com/andrewkolos/_flavors_test_world/tree/main/flavor_specific_assets

If I recall correctly, this app is configured to work with flavors on iOS, macOS, and Android.

Keep in mind this feature is not available on the `stable` branch/channel yet, so you'll have to make sure your flutter checkout is on `master` for it to work.

juarezfranco commented:
Finally on the stable branch!

giaur500 commented:
@andrewkolos 

Can I handle translated strings the same way? My app uses translated strings this way: https://docs.flutter.dev/ui/accessibility-and-internationalization/internationalization and I need to have separate .arb files for each flavor. But .arb are not resources, is there any way to do that?

juarezfranco commented:
@giaur500

I think it's crazy to create translations by flavor; for example, 2 flavors and 3 languages results in 6 generally large files to maintain, especially when the brand name changes almost every time.

I do like this

```
"welcome": "Welcome to {flavor, select, banana{Banana} apple{Apple} other{{flavor}}} Law Firm",
"@pronoun": {
  "placeholders": {
    "flavor": {
      "type": "String"
    }
  }
}
```

nicobritos commented:
Hi! Great work!! Is this feature present at the moment? In the Flutter 3.19.0 changelog I see that this commit has been reverted with https://github.com/flutter/flutter/pull/139787 but I am not sure why 🤔

giaur500 commented:
Indeed.. it's reverted, why?

andrewkolos commented:
I think the feature was relanded a few days after the revert. See https://github.com/flutter/flutter/pull/139834. I just tested the feature locally using the flutter stable branch and it works. Is this not working in 3.19.0?

giaur500 commented:
I haven't tried yet, my Flutter version is 3.19.2 stable. Regarding to revert, I'm asking to confirm if this  feature is not removed from some reason

agladyshev commented:
Works for me with Flutter 3.19.4. My .ipa is 46.5MB, .aab is 91.2MB. Android should have additional 41.2MB of assets. Seems about right.

giaur500 commented:
I think your case it not related to flavors. Flavors only exists on mobile platforms. If you don't want to pack files into apk or ipa, don't use them as resources, put them to some external directory instead

fenglintianxia6 commented:
> I think your case it not related to flavors. Flavors only exists on mobile platforms. If you don't want to pack files into apk or ipa, don't use them as resources, put them to some external directory instead

I saw @andrewkolos wrote something about platform. And the comment above just said he used this feature and solved his problem like platform-specific assets.  

and can you share more information about the `put them to some external directory instead`? how can I solve my problem with it

giaur500 commented:
`I saw @andrewkolos wrote something about platform.` - ok, let him to rise your question.  `and can you share more information about the put them to some external directory instead` -  I mean you can create directory on device file system and load files from there. Is there any reason why you want to bundle them as resources for web and desktop? BTW, on desktop or web nothing will be bundled, even resources reside in separate dir.

fenglintianxia6 commented:
I see. I think I know how to solve my problem. thanks.

LoadJulz commented:
I have different environments even on Web and overcome the missing flavor issue by just providing different main.dart entrypoints. Still an issue is that all my assets are loaded in Flutter Web even I don't need them in my build for that specific environment. Are there any solutions to this?

gnprice commented:
@LoadJulz That would be a good question to ask as a new issue. Generally that's a more effective venue for discussion than on the thread of an old merged PR. (You can always link to the old PR when it's relevant context.)

feduke-nukem commented:
Does that support flavors within separate packages?

For example:
```
|----lib
|-----|
|-----|----app <-- uses "image_kit"
|
|-----packages
|--------|
|--------|----image_kit <-- uses different assets in widget based on flavor
```

I have tried that already but it seems that it is not supported
